<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Setup" data-toc-modified-id="Setup-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Setup</a></span></li><li><span><a href="#Information-Retrieval" data-toc-modified-id="Information-Retrieval-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Information Retrieval</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Practical-Example" data-toc-modified-id="Practical-Example-2.0.1"><span class="toc-item-num">2.0.1&nbsp;&nbsp;</span>Practical Example</a></span></li><li><span><a href="#Boolean-Search-Queries" data-toc-modified-id="Boolean-Search-Queries-2.0.2"><span class="toc-item-num">2.0.2&nbsp;&nbsp;</span>Boolean Search Queries</a></span></li><li><span><a href="#Examples-of-Documents-That-Would-Satisfy-the-Above-Document-Term-Matrix" data-toc-modified-id="Examples-of-Documents-That-Would-Satisfy-the-Above-Document-Term-Matrix-2.0.3"><span class="toc-item-num">2.0.3&nbsp;&nbsp;</span>Examples of Documents That Would Satisfy the Above Document-Term Matrix</a></span><ul class="toc-item"><li><span><a href="#Query-Example-(nuclear-AND-treaty)" data-toc-modified-id="Query-Example-(nuclear-AND-treaty)-2.0.3.1"><span class="toc-item-num">2.0.3.1&nbsp;&nbsp;</span>Query Example <code>(nuclear AND treaty)</code></a></span></li><li><span><a href="#Query-Example-(nuclear-AND-treaty)-OR-((NOT-treaty)-AND-(nonproliferation-OR-Iran))" data-toc-modified-id="Query-Example-(nuclear-AND-treaty)-OR-((NOT-treaty)-AND-(nonproliferation-OR-Iran))-2.0.3.2"><span class="toc-item-num">2.0.3.2&nbsp;&nbsp;</span>Query Example <code>(nuclear AND treaty) OR ((NOT treaty) AND (nonproliferation OR Iran))</code></a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Tokenizing-Sentences" data-toc-modified-id="Tokenizing-Sentences-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Tokenizing Sentences</a></span><ul class="toc-item"><li><span><a href="#Definition-of-a-Token" data-toc-modified-id="Definition-of-a-Token-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Definition of a Token</a></span></li><li><span><a href="#Examples-of-Tokens/Words-Not-Being-the-Identical-Concepts" data-toc-modified-id="Examples-of-Tokens/Words-Not-Being-the-Identical-Concepts-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Examples of Tokens/Words Not Being the Identical Concepts</a></span></li><li><span><a href="#Definition-of-a-Document" data-toc-modified-id="Definition-of-a-Document-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Definition of a Document</a></span></li><li><span><a href="#Why-Is-Tokenization-Hard?" data-toc-modified-id="Why-Is-Tokenization-Hard?-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Why Is Tokenization Hard?</a></span><ul class="toc-item"><li><span><a href="#Sentence-Boundary-Detection" data-toc-modified-id="Sentence-Boundary-Detection-3.4.1"><span class="toc-item-num">3.4.1&nbsp;&nbsp;</span>Sentence Boundary Detection</a></span></li><li><span><a href="#Approaches-to-Sentence-Boundary-Detection" data-toc-modified-id="Approaches-to-Sentence-Boundary-Detection-3.4.2"><span class="toc-item-num">3.4.2&nbsp;&nbsp;</span>Approaches to Sentence Boundary Detection</a></span><ul class="toc-item"><li><span><a href="#Rules-Based" data-toc-modified-id="Rules-Based-3.4.2.1"><span class="toc-item-num">3.4.2.1&nbsp;&nbsp;</span>Rules-Based</a></span></li><li><span><a href="#Supervised/Unsupervised-Learning" data-toc-modified-id="Supervised/Unsupervised-Learning-3.4.2.2"><span class="toc-item-num">3.4.2.2&nbsp;&nbsp;</span>Supervised/Unsupervised Learning</a></span></li></ul></li><li><span><a href="#NLTK's-PunktSentenceTokenizer" data-toc-modified-id="NLTK's-PunktSentenceTokenizer-3.4.3"><span class="toc-item-num">3.4.3&nbsp;&nbsp;</span>NLTK's <code>PunktSentenceTokenizer</code></a></span></li></ul></li><li><span><a href="#Stemming" data-toc-modified-id="Stemming-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Stemming</a></span></li><li><span><a href="#Lemmatization" data-toc-modified-id="Lemmatization-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Lemmatization</a></span><ul class="toc-item"><li><span><a href="#Why-Stemming?" data-toc-modified-id="Why-Stemming?-3.6.1"><span class="toc-item-num">3.6.1&nbsp;&nbsp;</span>Why Stemming?</a></span></li><li><span><a href="#Differences-Between-Stemming-and-Lemmatization" data-toc-modified-id="Differences-Between-Stemming-and-Lemmatization-3.6.2"><span class="toc-item-num">3.6.2&nbsp;&nbsp;</span>Differences Between Stemming and Lemmatization</a></span></li></ul></li><li><span><a href="#Scoring-Metrics" data-toc-modified-id="Scoring-Metrics-3.7"><span class="toc-item-num">3.7&nbsp;&nbsp;</span>Scoring Metrics</a></span><ul class="toc-item"><li><span><a href="#Precision/Recall" data-toc-modified-id="Precision/Recall-3.7.1"><span class="toc-item-num">3.7.1&nbsp;&nbsp;</span>Precision/Recall</a></span></li><li><span><a href="#F1-Score" data-toc-modified-id="F1-Score-3.7.2"><span class="toc-item-num">3.7.2&nbsp;&nbsp;</span>F1 Score</a></span></li><li><span><a href="#Exercises" data-toc-modified-id="Exercises-3.7.3"><span class="toc-item-num">3.7.3&nbsp;&nbsp;</span>Exercises</a></span></li></ul></li><li><span><a href="#Removing-Stopwords" data-toc-modified-id="Removing-Stopwords-3.8"><span class="toc-item-num">3.8&nbsp;&nbsp;</span>Removing Stopwords</a></span><ul class="toc-item"><li><span><a href="#CountVectorize-the-Documents" data-toc-modified-id="CountVectorize-the-Documents-3.8.1"><span class="toc-item-num">3.8.1&nbsp;&nbsp;</span>CountVectorize the Documents</a></span></li></ul></li><li><span><a href="#When-You-Use-Lemmatization,-You-Need-to-Pass-in-Part-of-Speech" data-toc-modified-id="When-You-Use-Lemmatization,-You-Need-to-Pass-in-Part-of-Speech-3.9"><span class="toc-item-num">3.9&nbsp;&nbsp;</span>When You Use Lemmatization, You Need to Pass in Part of Speech</a></span><ul class="toc-item"><li><span><a href="#When-Should-You-Avoid-Removing-Stopwords?" data-toc-modified-id="When-Should-You-Avoid-Removing-Stopwords?-3.9.1"><span class="toc-item-num">3.9.1&nbsp;&nbsp;</span>When Should You Avoid Removing Stopwords?</a></span></li></ul></li><li><span><a href="#Exercise:" data-toc-modified-id="Exercise:-3.10"><span class="toc-item-num">3.10&nbsp;&nbsp;</span>Exercise:</a></span></li></ul></li><li><span><a href="#Vectorization-Techniques" data-toc-modified-id="Vectorization-Techniques-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Vectorization Techniques</a></span><ul class="toc-item"><li><span><a href="#Count-Vectorization" data-toc-modified-id="Count-Vectorization-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Count Vectorization</a></span></li></ul></li></ul></div>

# Setup

You'll need to download/install `nltk`, and then also download several of `nltk`'s modules.

In [5]:
# !conda install nltk

In [1]:
import nltk
nltk.download('punkt') # A popular NLTK sentence tokenizer
nltk.download('stopwords') # library of common English stopwords

[nltk_data] Downloading package punkt to
[nltk_data]     /Users/antheayang/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/antheayang/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

# Information Retrieval

> **Information retrieval (IR)** is finding material (usually documents) of an unstructured nature (usually text) that **satisfies an information need** from within large collections (usually stored on computers).

### Practical Example

A mobile app contains a gallery of about 8,000 pieces of artwork that they'd like to sell to users. Each artwork contains artwork name, artist name, descriptions, and some keyword tags. Th

### Boolean Search Queries

Example of a **document-term matrix**, from [Northeastern University, Fall 2006 Information Retrieval](http://www.ccs.neu.edu/home/jaa/CSG339.06F/Lectures/boolean.pdf):

<img src="https://raw.githubusercontent.com/ychennay/dso-560-nlp-text-analytics/main/images/boolean_search_query.png" alt="Different Stemming Techniques" style="width:600px;"/>

<div class='alert-success'>
    docs are D1, D2, ... while terms are nuclear, treaty,...
</div>

### Examples of Documents That Would Satisfy the Above Document-Term Matrix

* **Document 1**

"The new cookbook was a huge hit with younger millenials."
* **Document 2**

"Iran has significant nuclear production capabilities."
* **Document 3**

"The treaty that ended the American Revolution was signed soon thereafter."

#### Query Example `(nuclear AND treaty)`

`D7` would get returned since it is the only document that contains both `nuclear` and `treaty`.

#### Query Example `(nuclear AND treaty) OR ((NOT treaty) AND (nonproliferation OR Iran))`

* `D7`: satisfies LHS (left-hand side of expression) - contains `nuclear` and `treaty`
* `D5`: satisfied RHS - contains `nonproliferation` and not `treaty`)
* `D2`: satisfies RHS - contains `Iran` and not `treaty`)

<div class='alert-success'>
    vectorization: turn raw texts (i.e. docs) into doc-term matrix, i.e. numbers
</div>

# Tokenizing Sentences

## Definition of a Token

> A token is an instance of a sequence of characters in some particular document that are grouped together as a useful semantic unit for processing. ([Stanford NLP](https://nlp.stanford.edu/IR-book/html/htmledition/tokenization-1.html)):

<img src="https://raw.githubusercontent.com/ychennay/dso-560-nlp-text-analytics/main/images/tokenization.png" alt="Different Stemming Techniques" style="width:600px;"/>


## Examples of Tokens/Words Not Being the Identical Concepts

* Multi-word entities: `San Francisco`, `New York City`
* Phone numbers / dates: `(800) 123-4564` and `Apr. 30th, 2020`


At first, the task of tokenizing seems simple:
1. Split apart `corpus` into `documents`.
2. Split apart `documents` into `tokens`.

## Definition of a Document

A document is a distinct group of tokens. For example, in the Amazon toy product review dataset, each review can be considered a document. In *Tale of Two Cities*, a document could be either a sentence, or a paragraph, or a chapter - it all depends on the task at hand.

* **Sentiment Analysis**: each document is a review/comment/tweet
* **Authorship Classification** (determining if a piece of text was written by Person A or Person B): the entire text is a document

## Why Is Tokenization Hard?

### Sentence Boundary Detection

Deciding where one sentence ends and another sentence begins is easy for humans, but extremely complex for machines. A prime cause of this is ambiguity of punctuation marks. Particularly in English, periods can mean many things beyond denoting the end of a sentence:
* abbreviations
* decimal points
* domain names and email addresses

Furthermore, other "obvious" sentence boundaries like question marks and exclamation marks are becoming increasingly ambiguous due to slang, emoticons, or just emphasis (`"What are you doing???"` should not be split into 3 separate sentences).

### Approaches to Sentence Boundary Detection

#### Rules-Based

* Using regex: `((?<=[a-z0-9][.?!])|(?<=[a-z0-9][.?!]\"))(\s|\r\n)(?=\"?[A-Z])` (don't worry if you can't understand this yet- it includes usage of several advanced techniques such as **lookahead/behind references** which we have not covered).

#### Supervised/Unsupervised Learning

* Maximum entropy models ([UC Berkeley's SATZ Adaptive Sentence Boundary Detector](https://web.archive.org/web/20070922132340/http://elib.cs.berkeley.edu/src/satz/))
* NLTK's `Punkt` sentence tokenizer


<div class='alert-success'>
    NLTK uses a pre-trained supervised model to detect sentences
</div>

In [3]:
# Why not just tokenize myself?
text = "I made two purchases today! I bought a bag of grapes for $4.99, \
but then... realized John Francis already bought some at the Y.M.C.A!"

In [3]:
# trying to write our own tokenizer
text.split(".")

['I made two purchases today! I bought a bag of grapes for $4',
 '99, but then',
 '',
 '',
 ' realized John Francis already bought some at the Y',
 'M',
 'C',
 'A!']

What are some of the issues you notice with the above approach?

### NLTK's `PunktSentenceTokenizer`

> `PunktSentenceTokenizer` is an **sentence boundary detection algorithm** that must be trained to be used. NLTK already includes a pre-trained version of the `PunktSentenceTokenizer` ([StackOverflow](https://stackoverflow.com/questions/35275001/use-of-punktsentencetokenizer-in-nltk))

In [4]:
# Using NLTK sent_tokenize()
sent_text = nltk.sent_tokenize(text) # this gives us a list of sentences
sent_text

['I made two purchases today!',
 'I bought a bag of grapes for $4.99, but then... realized John Francis already bought some at the Y.M.C.A!']

In [9]:
tokens = nltk.word_tokenize(text)
stopword_removed = [token for token in tokens if token not in stopwords]
stopword_removed

TypeError: argument of type 'WordListCorpusReader' is not iterable

<div class='alert-danger'>
    bugs!!!
    </div>

## Stemming

Stemming is the process of reducing inflection in words to their root forms such as mapping a group of words to the same stem even if the stem itself is not a valid word in the language [Source](https://www.datacamp.com/community/tutorials/stemming-lemmatization-python)

<img src="../images/stemming-examples.png" alt="Different Stemming Techniques" style="width:600px;"/>

In Python, we can use **`nltk.stem.porter.PorterStemmer`** stem our words:

```python
stemmer = PorterStemmer()
print(stemmer.stem("caressed"))  # caress
print(stemmer.stem("athlete"))  # athlet
print(stemmer.stem("athletics"))  # athlet
print(stemmer.stem("media"))  # media
print(stemmer.stem("photography"))  # photographi
print(stemmer.stem("sexy"))  # sexi
print(stemmer.stem("journalling"))  # journal
print(stemmer.stem("Slovakia")) # slovakia
print(stemmer.stem("corpora")) # corpora
print(stemmer.stem("thieves")) # thiev
print(stemmer.stem("rocks")) # rock
```

In [9]:
from nltk.stem.porter import PorterStemmer
stemmer = PorterStemmer()
stemmer.stem("photagraphy")

'photagraphi'

<div class='alert-success'>
    stemmer reduces words to their root form, useful for word count, yet sacrifice the meanings
</div>

In [7]:
from nltk.stem.porter import PorterStemmer
stemmer = PorterStemmer()
stemmer.stem("photo"), stemmer.stem("photography"), stemmer.stem("photagraphy")

('photo', 'photographi', 'photagraphi')

<div class='alert-success'>
    it may have some weaknesses and does not account for typo
</div>

## Lemmatization

Lemmatization is closely related to stemming. However, while stemming looks only at the individual word itself, and considers the usage of the word (ie. part of speech, is this word a noun, a verb, etc.). For example, if we compared the [outputs of stemming and lemmatization certain ambiguous French words](https://blog.bitext.com/lemmatization-vs-stemming):

<img src="https://raw.githubusercontent.com/ychennay/dso-560-nlp-text-analytics/main/images/comparison.png" alt="Different Stemming Techniques" style="width:600px;"/>

```python
from nltk.stem import WordNetLemmatizer
lemmatizer = WordNetLemmatizer()
print(lemmatizer.lemmatize("caressed")) #caressed
print(lemmatizer.lemmatize("athlete")) #athlete
print(lemmatizer.lemmatize("athletics")) #athletics
print(lemmatizer.lemmatize("media"))
print(lemmatizer.lemmatize("photography")) #photography
print(lemmatizer.lemmatize("sexy")) #sexy
print(lemmatizer.lemmatize("journalling")) #journalling
print(lemmatizer.lemmatize("Slovakia")) #Slovakia
print(lemmatizer.lemmatize("corpora")) # corpus
print(lemmatizer.lemmatize("thieves")) # thief
print(lemmatizer.lemmatize("rocks")) #rock
```

<div class='alert-success'>
    refers to 3.9 for the usage of lemmatization. do NOT just lemmatize("xxx"), which treats xxx as noun. lemmatize words with speech.
</div>

### Why Stemming?
- smaller and faster
- simplicity in "good enough"
- can often **provide higher recall (coverage)** if you are using it for text searching: `drives` and `drivers` will likely shorten to `driv`, which may be useful if your search engine wants to make sure to get all relevant documents, even at the cost of surfacing a few irrelevant documents
- could potentially be more useful for predictive models that tend to overfit

### Differences Between Stemming and Lemmatization
> **Stemming** usually refers to a crude heuristic process that chops off the ends of words in the hope of achieving this goal correctly most of the time, and often includes the removal of derivational affixes. **Lemmatization** usually refers to doing things properly with the use of a vocabulary and morphological analysis of words, normally aiming to remove inflectional endings only and to return the base or dictionary form of a word, which is known as the lemma . If confronted with the token `saw`, stemming might return just `s`, whereas lemmatization would attempt to return either `see` or `saw` depending on whether the use of the token was as a verb or a noun. ([Stanford NLP Group](https://nlp.stanford.edu/IR-book/html/htmledition/stemming-and-lemmatization-1.html)) 

<div class='alert-success'>
    <p>stemming leads to higher recall, and lower precision
    <p>lemmatization also leads to higher recall, it's just nuance stemming
</div>

## Scoring Metrics

<img src="https://raw.githubusercontent.com/ychennay/dso-560-nlp-text-analytics/main/images/confusion_matrix2.png" alt="Different Stemming Techniques" style="width:600px;"/>

### Precision/Recall

**Recall:** What percent of the positive classes did the model successfully predict?
**Precision:** When a model predicted a positive class, what percentage of the time was it correct?

In terms of NLP / stemming / lemmatization:

**Recall**: After processing (tokenizing, stemming/lemmatizing) the data, what percent of the relevant search results were surfaced? Ie. - when a user searches for `blue jeans`, did all the results returned include all the relevant items (blue-ish colored denim pants)?

**Precision**: After processing (tokenizing, stemming/lemmatizing) the data, what percent of the results returned were relevant?

<img src="https://raw.githubusercontent.com/ychennay/dso-560-nlp-text-analytics/main/images/matrix_practice3.png" alt="Different Stemming Techniques" style="width:600px;"/>


<div class='alert-success'>
    search engines usually prefer <i>precision</i>, as there're so many relevants
</div>

### F1 Score

The F1 score of a model represents the harmonic mean between precision and recall, and is defined as 

$$
\begin{equation}
F_{1} = 2 * \frac{P * R}{P + R}
\end{equation}
$$

An F1 score is often a good measure 
* our dataset target is class imbalanced (ie. 96% positive, 4% negative)
* when we want to balance optimizing for both precision and recall

### Exercises

Calculate:

1. Overall **accuracy**: $\frac{?}{?}$

If $C$ is the number of correct predictions, and $N$ is the number of total samples:
\begin{equation}
a = \frac{C}{N}
\end{equation}

\begin{equation}
a = \frac{30}{36}
\end{equation}

\begin{equation}
a = 0.833
\end{equation}

2. **Precision:** $\frac{?}{?}$

If $TP$ is the number of true positives, and $FP$ is the number of false positives:
\begin{equation}
P = \frac{TP}{TP + FP}
\end{equation}

\begin{equation}
P = \frac{10}{10 + 2}
\end{equation}

\begin{equation}
P = 0.833
\end{equation}

3. **Recall:** $\frac{?}{?}$

If $TP$ is the number of true positives, and $FN$ is the number of false negatives:
\begin{equation}
R = \frac{TP}{TP + FN}
\end{equation}

\begin{equation}
R = \frac{10}{10 + 4}
\end{equation}

\begin{equation}
R = 0.714
\end{equation}

4. **F1 Score**

\begin{equation}
F_{1} = 2 * \frac{P * R}{P + R}
\end{equation}

\begin{equation}
F_{1} = 2 * \frac{0.833 * 0.714}{0.833 + 0.714}
\end{equation}

\begin{equation}
F_{1} =0.769
\end{equation}

## Removing Stopwords

It's your call ultimately if you want to remove stopwords. There are advantages and disadvantages to both approaches. You will first need to run `nltk.download("stopwords")` to download the set of stopwords for NLTK:

In [8]:
from nltk.corpus import stopwords
print(set(stopwords.words('english'))) # see the set of words NLTK considers stopwords

{'what', 'being', 'isn', "wasn't", 'is', 'them', 'above', "won't", 'so', 'hadn', 'more', 'these', 'we', "hadn't", 'mightn', 'will', 'myself', 'she', 's', 'can', 'not', 'their', 'it', 'was', 'no', "don't", "you'll", 'how', 'should', 'as', 'until', 'such', 'll', 'during', 'didn', 'before', 'of', 'themselves', 'own', 'between', 'once', 'out', 'couldn', 'or', 'by', 'don', "that'll", "haven't", 'through', 'too', 'now', 'this', 'd', 'over', 'in', 'when', 'do', "she's", 'doesn', 'they', 'because', "should've", "you're", 'that', "couldn't", 'shouldn', "doesn't", "mustn't", 'yourselves', 'off', 'very', 'hers', "wouldn't", 'am', 'its', 'me', 'his', 'below', 'our', 'then', 'been', 'he', "needn't", 'himself', "aren't", 'both', 'about', 'herself', 'with', 'few', "shan't", 'again', 'and', 'other', 'into', 'on', 'same', 'my', 'won', 'your', "mightn't", 'from', 'haven', "hasn't", 'doing', 'a', 'nor', 'you', 'weren', 'each', 'those', 't', 'needn', 'all', 'where', 'why', 'has', 'theirs', 'down', 'had', 

In [10]:
stopwords = set(stopwords.words('english'))
"the" in stopwords

True

In [12]:
import pandas as pd
import numpy as np
# import in a dataset of baltimore's public art galleries
public_art_df: pd.DataFrame = pd.read_csv("../datasets/baltimore_public_art.csv")

public_art_df = public_art_df.replace(np.nan, '', regex=True)
public_art_df.head()
# use the titleOfArtwork field
titles_of_artworks: pd.Series = public_art_df["titleOfArtwork"]

In [16]:
titles_of_artworks

0                                           Family Group
1                               Moses and Christ Reliefs
2                                William Donald Schaefer
3      Atlantic Blue Roller Column or Interlocking Pi...
4                                        Baltimore Pylon
                             ...                        
685                                                     
686                                   Birds Nest Balcony
687                                         Male/ Female
688                                              silence
689                                        Inertia Study
Name: titleOfArtwork, Length: 690, dtype: object

### CountVectorize the Documents

In [20]:
from sklearn.feature_extraction.text import CountVectorizer
# vectorizer = CountVectorizer(stopwords="english")
vectorizer = CountVectorizer()
X = vectorizer.fit_transform(titles_of_artworks)
X

<690x619 sparse matrix of type '<class 'numpy.int64'>'
	with 1169 stored elements in Compressed Sparse Row format>

<div class='alert-success'>
    690 docs, 619 tokens
</div>

<div class='alert-success'>
    <p><i>fit_transform:</i> 
    <p>fit has already fitted the object to the data and created certain attributes
    <p>transform would transform the data to e.g. numpy array
</div>

In [21]:
X = X.toarray()
X

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [22]:
# corpus_df = pd.DataFrame(X, columns=vectorizer.get_feature_names()) # deprecated
corpus_df = pd.DataFrame(X, columns=vectorizer.get_feature_names_out())
corpus_df

10  1890  1912  1940  1984  31  420  43  aegean  african  ...  william  \
0     0     0     0     0     0   0    0   0       0        0  ...        0   
1     0     0     0     0     0   0    0   0       0        0  ...        0   
2     0     0     0     0     0   0    0   0       0        0  ...        1   
3     0     0     0     0     0   0    0   0       0        0  ...        0   
4     0     0     0     0     0   0    0   0       0        0  ...        0   
..   ..   ...   ...   ...   ...  ..  ...  ..     ...      ...  ...      ...   
685   0     0     0     0     0   0    0   0       0        0  ...        0   
686   0     0     0     0     0   0    0   0       0        0  ...        0   
687   0     0     0     0     0   0    0   0       0        0  ...        0   
688   0     0     0     0     0   0    0   0       0        0  ...        0   
689   0     0     0     0     0   0    0   0       0        0  ...        0   

     wishbone  with  woman  women  world  young  yuai  yum  zappa  
0           0     0      0      0      0      0     0    0      0  
1           0     0      0      0      0      0     0    0      0  
2           0     0      0      0      0      0     0    0      0  
3           0     0      0      0      0      0     0    0      0  
4           0     0      0      0      0      0     0    0      0  
..        ...   ...    ...    ...    ...    ...   ...  ...    ...  
685         0     0      0      0      0      0     0    0      0  
686         0     0      0      0      0      0     0    0      0  
687         0     0      0      0      0      0     0    0      0  
688         0     0      0      0      0      0     0    0      0  
689         0     0      0      0      0      0     0    0      0  

[690 rows x 619 columns]

In [ ]:
corpus_df

In [ ]:
# iterate through the Pandas dataframe, and drop the columns that reflect stopwords:
original_columns = corpus_df.columns # get existing columns

to_drop_columns = set(original_columns).intersection(set(stopwords)) # get the list of words to drop
to_drop_columns

In [ ]:
from nltk.stem import WordNetLemmatizer
lemmatizer = WordNetLemmatizer()
lemmatizer.lemmatize("liked", pos="v")

In [ ]:
from nltk.stem import WordNetLemmatizer
lemmatizer = WordNetLemmatizer()
lemmatizer.lemmatize("liked") # assumes this is a noun

## When You Use Lemmatization, You Need to Pass in Part of Speech

In [ ]:
## lemmatization
## https://gist.github.com/gaurav5430/9fce93759eb2f6b1697883c3782f30de#file-nltk-lemmatize-sentences-py
import nltk
from nltk.stem import WordNetLemmatizer
from nltk.corpus import wordnet

lemmatizer = WordNetLemmatizer()

# function to convert nltk tag to wordnet tag
def nltk_tag_to_wordnet_tag(nltk_tag):
    if nltk_tag.startswith('J'):
        return wordnet.ADJ
    elif nltk_tag.startswith('V'):
        return wordnet.VERB
    elif nltk_tag.startswith('N'):
        return wordnet.NOUN
    elif nltk_tag.startswith('R'):
        return wordnet.ADV
    else:          
        return None

def lemmatize_sentence(sentence):
    #tokenize the sentence and find the POS tag for each token
    nltk_tagged = nltk.pos_tag(nltk.word_tokenize(sentence))  
    #tuple of (token, wordnet_tag)
    wordnet_tagged = map(lambda x: (x[0], nltk_tag_to_wordnet_tag(x[1])), nltk_tagged)
    lemmatized_sentence = []
    for word, tag in wordnet_tagged:
        if tag is None:
            #if there is no available tag, append the token as is
            lemmatized_sentence.append(word)
        else:        
            #else use the tag to lemmatize the token
            lemmatized_sentence.append(lemmatizer.lemmatize(word, tag))
    return " ".join(lemmatized_sentence)

In [ ]:
print(f"Dataframe shape was {corpus_df.shape}")
corpus_df.drop(columns=to_drop_columns, inplace=True)
print(f"Dataframe shape is now{corpus_df.shape}")

### When Should You Avoid Removing Stopwords?

The below example is taken from a Medium article by Gagandeep Singh, ["Why you should avoid removing STOPWORDS"](https://towardsdatascience.com/why-you-should-avoid-removing-stopwords-aa7a353d2a52).

In [ ]:
reviews = [
    ("The product is really very good", "Positive"),
    ("The products seems to be good", "Positive"),
    ("Good product. I really liked it", "Positive"),
    ("I didn’t like the product", "Negative"),
    ("The product is not good.", "Negative")
]

In [ ]:
cleaned_reviews = []
nltk_stopwords = set(stopwords.words('english'))

# This iterates through each of the reviews, splitting the review into distinct tokens
# Then it checks each token for whether or not it is a stopword, before adding them back into a "cleaned_review"
for review in reviews:
    words = nltk.word_tokenize(review[0])
    new_words = []
    for word in words:
        if word.lower() in nltk_stopwords:
            continue
        new_words.append(word)
    cleaned_review = " ".join(new_words)
    cleaned_reviews.append((cleaned_review, review[1]))

Singh:
> If you are working with **basic NLP techniques like BOW, Count Vectorizer or TF-IDF(Term Frequency and Inverse Document Frequency)** then removing stopwords is a good idea because stopwords act like noise for these methods. If you working with **LSTMs or other models which capture the semantic meaning and the meaning of a word depends on the context of the previous text**, then it becomes important not to remove stopwords.

## Exercise:

##### 1. For each of the following statements, label them True or False. If False, briefly explain why:

A. *Text typically should be processed via either stemming or lemmatization, but not both.*

B. *Texts processed using lemmatization will typically have higher recall than stemming.*

C. *If the **F1 score** of a model is **1.0 (100%)**, then the accuracy of your model must also be **100%**.*

D. *Stemming **increases the size of the vocabulary** (the vocabulary is the set of all tokens found inside the corpus)*

##### 2. Calculate precison and recall given the following results from a confusion matrix:

<img src="../images/exercise.jpeg" alt="Different Stemming Techniques" style="width:600px;"/>

##### 3. Provide an example of how stemming can improve recall.

##### 4. Provide an example of when stemming might reduce precision.

# Vectorization Techniques

## Count Vectorization
<img src="https://raw.githubusercontent.com/ychennay/dso-560-nlp-text-analytics/main/images/count_vectorizer.png" alt="Different Stemming Techniques" style="width:600px;"/>

You can use **`sklearn.feature_extraction.text.CountVectorizer`** to easily convert your corpus into a bag of words matrix:

```python
from sklearn.feature_extraction.text import CountVectorizer
vectorizer = CountVectorizer()
data_corpus = ["John likes to watch movies. Mary likes movies too.", 
"John also likes to watch football games. Mary does not like football much."]
X = vectorizer.fit_transform(data_corpus) 
```
Note that the output `X` here is not your traditional Numpy matrix! Calling **`type(X)`** here will yield **`<class 'scipy.sparse.csr.csr_matrix'>`**, which is a **CSR ([compressed sparse row format matrix](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.sparse.csr_matrix.html))**. To convert it into an actual matrix, call the `toarray()` method:

```python
X.toarray()
```
Your output will be 

```
array([[0, 0, 0, 0, 1, 0, 2, 1, 2, 0, 0, 1, 1, 1],
       [1, 1, 2, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1]], dtype=int64)
```
Notice that using **`X.shape`** $\rightarrow$ `(2,14)`, indicating a total vocabulary size $V$ of 14. To get what word each of the 14 columns corresponds to, use **`vectorizer.get_feature_names()`**:
```
['also', 'does', 'football', 'games', 'john', 'like', 'likes', 'mary', 'movies', 'much', 'not', 'to', 'too', 'watch']
```

Notice, however, that as the vocabulary size $V$ increases, the percent of the matrix taken up by zero values increases:

```python
corpus = [
    "Some analysts think demand could drop this year because a large number of homeowners take on remodeling projectsafter buying a new property. With fewer homes selling, home values easing, and mortgage rates rising, they predict home renovations could fall to their lowest levels in three years.", 
    
          "Most home improvement stocks are expected to report fourth-quarter earnings next month.",
    
         "The conversation boils down to how much leverage management can get out of its wide-ranging efforts to re-energize operations, branding, digital capabilities, and the menu–and, for investors, how much to pay for that.",
    
    "RMD’s software acquisitions, efficiency, and mix overcame pricing and its gross margin improved by 90 bps Y/Y while its operating margin (including amortization) improved by 80 bps Y/Y. Since RMD expects the slower international flow generator growth to continue for the next few quarters, we have lowered our organic growth estimates to the mid-single digits."
]

X = vectorizer.fit_transform(corpus).toarray() 
```


In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
# list of text documents
text = ["It's still early, so box-office disappointments are still among the highest-grossing movies of the year.", 
        "That movie was terrific", "You love cats", 
        "Pay for top executives at big US companies is vastly higher than what everyday workers make, and a new report from The Wall Street Journal has found that CEOs have hit an eye-popping milestone in the size of their monthly paychecks."]
# create the transform
vectorizer = CountVectorizer()

# tokenize and build vocab
vectorizer.fit(text)

# vectorize the corpus
vector = vectorizer.transform(text)

# summarize encoded vector
print(vector.shape)

# Notice what type of object this is
print(type(vector))

In [ ]:
vectorizer = CountVectorizer()
docs = ["I love cats", "cats love I"]

vectorizer.fit_transform(docs).toarray()

In [ ]:
# see the outputted vectors
print(vector.toarray())
print(vectorizer.get_feature_names())

In [ ]:
# load vectorized corpus into Pandas dataframe
import pandas as pd
corpus_df = pd.DataFrame(vector.toarray(), columns=vectorizer.get_feature_names())
corpus_df.describe()